File: Source/Server/Game/Objects/Moral.vb
Imports Core
Imports Mirage.Sharp.Asfw
Imports Newtonsoft.Json
Imports Newtonsoft.Json.Linq

Friend Module Moral

#Region "Database"
    Sub ClearMorals()
        Dim i As Integer

        ReDim Type.Moral(MAX_MORALS)

        For i = 1 To MAX_MORALS
            ClearMoral(i)
        Next
    End Sub

    Sub ClearMoral(moralNum As Integer)
        Type.Moral(moralNum).Name = ""
        Type.Moral(moralNum).Color = 0
        Type.Moral(moralNum).CanCast = 0
        Type.Moral(moralNum).CanDropItem = 0
        Type.Moral(moralNum).CanPK = 0
        Type.Moral(moralNum).CanPickupItem = 0
        Type.Moral(moralNum).CanUseItem = 0
        Type.Moral(moralNum).DropItems = 0
        Type.Moral(moralNum).LoseExp = 0
        Type.Moral(moralNum).NPCBlock = 0
        Type.Moral(moralNum).PlayerBlock = 0
    End Sub

    Sub LoadMoral(moralNum As Integer)
        Dim data As JObject

        data = SelectRow(moralNum, "moral", "data")

        If data Is Nothing Then
            ClearMoral(moralNum)
            Exit Sub
        End If

        Dim moralData = JObject.FromObject(data).ToObject(Of MoralStruct)()
        Type.Moral(moralNum) = moralData
    End Sub

    Sub LoadMorals()
        Dim i As Integer

        For i = 1 To MAX_MORALS
            LoadMoral(i)
        Next
    End Sub

    Sub SaveMoral(moralNum As Integer)
        Dim json As String = JsonConvert.SerializeObject(Type.Moral(moralNum)).ToString()

        If RowExists(moralNum, "moral") Then
            UpdateRow(moralNum, json, "moral", "data")
        Else
            InsertRow(moralNum, json, "moral")
        End If
    End Sub

    Sub SaveMorals()
        Dim i As Integer

        For i = 1 To MAX_MORALS
            SaveMoral(i)
        Next
    End Sub

    Function MoralData(moralNum As Integer) As Byte()
        Dim buffer As New ByteStream(4)

        buffer.WriteInt32(moralNum)
        buffer.WriteString(Type.Moral(moralNum).Name)
        buffer.WriteByte(Type.Moral(moralNum).Color)
        buffer.WriteBoolean(Type.Moral(moralNum).NPCBlock)
        buffer.WriteBoolean(Type.Moral(moralNum).PlayerBlock)
        buffer.WriteBoolean(Type.Moral(moralNum).DropItems)
        buffer.WriteBoolean(Type.Moral(moralNum).CanCast)
        buffer.WriteBoolean(Type.Moral(moralNum).CanDropItem)
        buffer.WriteBoolean(Type.Moral(moralNum).CanPickupItem)
        buffer.WriteBoolean(Type.Moral(moralNum).CanPK)
        buffer.WriteBoolean(Type.Moral(moralNum).DropItems)
        buffer.WriteBoolean(Type.Moral(moralNum).LoseExp)

        Return buffer.ToArray
    End Function

#End Region

#Region "Outgoing Packets"

    Sub SendMorals(index As Integer)
        Dim i As Integer

       For i = 1 To MAX_MORALS
            If Len(Type.Moral(i).Name) > 0 Then
                SendUpdateMoralTo(index, i)
            End If
        Next

    End Sub

    Sub SendUpdateMoralTo(index As Integer, moralNum As Integer)
        Dim buffer As ByteStream
        buffer = New ByteStream(4)
        buffer.WriteInt32(ServerPackets.SUpdateMoral)

        buffer.WriteBlock(MoralData(moralNum))

        Socket.SendDataTo(index, buffer.Data, buffer.Head)
        buffer.Dispose()
    End Sub

    Sub SendUpdateMoralToAll(moralNum As Integer)
        Dim buffer As ByteStream
        buffer = New ByteStream(4)
        buffer.WriteInt32(ServerPackets.SUpdateMoral)

        buffer.WriteBlock(MoralData(moralNum))

        SendDataToAll(buffer.Data, buffer.Head)
        buffer.Dispose()
    End Sub


#End Region

#Region "Incoming Packets"
    Sub Packet_RequestEditMoral(index As Integer, ByRef data() As Byte)
        Dim buffer = New ByteStream(4)

        If GetPlayerAccess(index) < AccessType.Developer Then Exit Sub
        If TempPlayer(index).Editor > 0 Then Exit Sub

        Dim user As String

        user = IsEditorLocked(index, EditorType.Moral)

        If user <> "" Then
            PlayerMsg(index, "The game editor is locked and being used by " + user + ".", ColorType.BrightRed)
            Exit Sub
        End If

        SendMorals(index)

        TempPlayer(index).Editor = EditorType.Moral

        buffer.WriteInt32(ServerPackets.SMoralEditor)
        Socket.SendDataTo(index, buffer.Data, buffer.Head)

        buffer.Dispose()

    End Sub

    Sub Packet_SaveMoral(index As Integer, ByRef data() As Byte)
        Dim moralNum As Integer, i As Integer
        Dim buffer As New ByteStream(data)

        ' Prevent hacking
        If GetPlayerAccess(index) < AccessType.Developer Then Exit Sub

        moralNum = buffer.ReadInt32()

        ' Prevent hacking
        If moralNum <= 0 Or moralNum > MAX_MORALS Then Exit Sub

        With Type.Moral(moralNum)
            .Name = buffer.ReadString()
            .Color = buffer.ReadByte()
            .CanCast = buffer.ReadBoolean()
            .CanPK = buffer.ReadBoolean()
            .CanDropItem = buffer.ReadBoolean()
            .CanPickupItem = buffer.ReadBoolean()
            .CanUseItem = buffer.ReadBoolean()
            .DropItems = buffer.ReadBoolean()
            .LoseExp = buffer.ReadBoolean()
            .PlayerBlock = buffer.ReadBoolean()
            .NPCBlock = buffer.ReadBoolean()
        End With

        ' Save it
        SendUpdateMoralToAll(moralNum)
        SaveMoral(moralNum)
        Addlog(GetPlayerLogin(index) & " saved moral #" & moralNum & ".", ADMIN_LOG)
        SendMorals(index)
    End Sub

    Sub Packet_RequestMoral(index As Integer, ByRef data() As Byte)
        SendMorals(index)
    End Sub
#End Region
End Module

File: Source/Server/Game/Objects/Moral.cs
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;
using Mirage.Sharp.Asfw;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using static Core.Packets;
using static Core.Type;
using static Core.Global.Command;

namespace Server
{

    public class Moral
    {

        #region Database
        public static void ClearMoral(int moralNum)
        {
            Core.Data.Moral[moralNum].Name = "";
            Core.Data.Moral[moralNum].Color = 0;
            Core.Data.Moral[moralNum].CanCast = false;
            Core.Data.Moral[moralNum].CanDropItem = false;
            Core.Data.Moral[moralNum].CanPK = false;
            Core.Data.Moral[moralNum].CanPickupItem = false;
            Core.Data.Moral[moralNum].CanUseItem = false;
            Core.Data.Moral[moralNum].DropItems = false;
            Core.Data.Moral[moralNum].LoseExp = false;
            Core.Data.Moral[moralNum].NpcBlock = false;
            Core.Data.Moral[moralNum].PlayerBlock = false;
        }

        public static async System.Threading.Tasks.Task LoadMoralAsync(int moralNum)
        {
            JObject data;

            data = await Database.SelectRowAsync(moralNum, "moral", "data");

            if (data is null)
            {
                ClearMoral(moralNum);
                return;
            }

            var moralData = JObject.FromObject(data).ToObject<Core.Type.Moral>();
            Core.Data.Moral[moralNum] = moralData;
        }

        public static async System.Threading.Tasks.Task LoadMoralsAsync()
        {
            int i;

            var loopTo = Core.Constant.MAX_MORALS;
            for (i = 0; i < loopTo; i++)
                await System.Threading.Tasks.Task.Run(() => LoadMoralAsync(i));
        }

        public static void SaveMoral(int moralNum)
        {
            string json = JsonConvert.SerializeObject(Core.Data.Moral[moralNum]).ToString();

            if (Database.RowExists(moralNum, "moral"))
            {
                Database.UpdateRow(moralNum, json, "moral", "data");
            }
            else
            {
                Database.InsertRow(moralNum, json, "moral");
            }
        }

        public static void SaveMorals()
        {
            int i;

            var loopTo = Core.Constant.MAX_MORALS;
            for (i = 0; i < loopTo; i++)
                SaveMoral(i);
        }

        public static byte[] MoralData(int moralNum)
        {
            var buffer = new ByteStream(4);

            buffer.WriteInt32(moralNum);
            buffer.WriteString(Core.Data.Moral[moralNum].Name);
            buffer.WriteByte(Core.Data.Moral[moralNum].Color);
            buffer.WriteBoolean(Core.Data.Moral[moralNum].NpcBlock);
            buffer.WriteBoolean(Core.Data.Moral[moralNum].PlayerBlock);
            buffer.WriteBoolean(Core.Data.Moral[moralNum].DropItems);
            buffer.WriteBoolean(Core.Data.Moral[moralNum].CanCast);
            buffer.WriteBoolean(Core.Data.Moral[moralNum].CanDropItem);
            buffer.WriteBoolean(Core.Data.Moral[moralNum].CanPickupItem);
            buffer.WriteBoolean(Core.Data.Moral[moralNum].CanPK);
            buffer.WriteBoolean(Core.Data.Moral[moralNum].DropItems);
            buffer.WriteBoolean(Core.Data.Moral[moralNum].LoseExp);

            return buffer.ToArray();
        }

        #endregion

        #region Outgoing Packets

        public static void SendMorals(int index)
        {
            int i;

            var loopTo = Core.Constant.MAX_MORALS;
            for (i = 0; i < loopTo; i++)
            {
                if (Strings.Len(Core.Data.Moral[i].Name) > 0)
                {
                    SendUpdateMoralTo(index, i);
                }
            }

        }

        public static void SendUpdateMoralTo(int index, int moralNum)
        {
            ByteStream buffer;
            buffer = new ByteStream(4);
            buffer.WriteInt32((int) ServerPackets.SUpdateMoral);

            buffer.WriteBlock(MoralData(moralNum));

            NetworkConfig.Socket.SendDataTo(index, buffer.UnreadData, buffer.WritePosition);
            buffer.Dispose();
        }

        public static void SendUpdateMoralToAll(int moralNum)
        {
            ByteStream buffer;
            buffer = new ByteStream(4);
            buffer.WriteInt32((int) ServerPackets.SUpdateMoral);

            buffer.WriteBlock(MoralData(moralNum));

            NetworkConfig.SendDataToAll(buffer.UnreadData, buffer.WritePosition);
            buffer.Dispose();
        }


        #endregion

        #region Incoming Packets
        public static void Packet_RequestEditMoral(int index, ref byte[] data)
        {
            var buffer = new ByteStream(4);

            if (GetPlayerAccess(index) < (byte) Core.AccessLevel.Developer)
                return;

            string user;

            user = IsEditorLocked(index, (byte) Core.EditorType.Moral);

            if (!string.IsNullOrEmpty(user))
            {
                NetworkSend.PlayerMsg(index, "The game editor is locked and being used by " + user + ".", (int) Core.Color.BrightRed);
                return;
            }

            SendMorals(index);

            Core.Data.TempPlayer[index].Editor = (byte) Core.EditorType.Moral;

            buffer.WriteInt32((int) ServerPackets.SMoralEditor);
            NetworkConfig.Socket.SendDataTo(index, buffer.UnreadData, buffer.WritePosition);

            buffer.Dispose();

        }

        public static void Packet_SaveMoral(int index, ref byte[] data)
        {
            int moralNum;
            int i;
            var buffer = new ByteStream(data);

            // Prevent hacking
            if (GetPlayerAccess(index) < (byte) Core.AccessLevel.Developer)
                return;

            moralNum = buffer.ReadInt32();

            // Prevent hacking
            if (moralNum < 0 | moralNum > Core.Constant.MAX_MORALS)
                return;

            {
                ref var withBlock = ref Core.Data.Moral[moralNum];
                withBlock.Name = buffer.ReadString();
                withBlock.Color = buffer.ReadByte();
                withBlock.CanCast = buffer.ReadBoolean();
                withBlock.CanPK = buffer.ReadBoolean();
                withBlock.CanDropItem = buffer.ReadBoolean();
                withBlock.CanPickupItem = buffer.ReadBoolean();
                withBlock.CanUseItem = buffer.ReadBoolean();
                withBlock.DropItems = buffer.ReadBoolean();
                withBlock.LoseExp = buffer.ReadBoolean();
                withBlock.PlayerBlock = buffer.ReadBoolean();
                withBlock.NpcBlock = buffer.ReadBoolean();
            }

            // Save it
            SendUpdateMoralToAll(moralNum);
            SaveMoral(moralNum);
            Core.Log.Add(GetAccountLogin(index) + " saved moral #" + moralNum + ".", Constant.ADMIN_LOG);
            SendMorals(index);
        }

        public static void Packet_RequestMoral(int index, ref byte[] data)
        {
            SendMorals(index);
        }
        #endregion
    }
}